lib/reply.js

```diff
@@ -104,7 +106,16 @@ Object.defineProperties(Reply.prototype, {
   }
 })
 
+Reply.prototype.hijack = function () {
+  this[kReplyHijacked] = true
+  return this
+}
+
 Reply.prototype.send = function (payload) {
+  if (this[kReplyHijacked] === true) {
```

Comment:
Might be good to add a comment or something like that here to explain what the heck this is for. Surprisingly simple change though!

Reply:
Do you think that it should be an error to `reply.send()` from user-land code after `reply.hijack()`ing?

Reply:
Oh also, can anyone say if there is some important `Reply` cleanup code that still needs to be run despite not being `sent`? Like removing it from any global maps or removing listeners or something?

Reply:
Will add a comment and tests for it.
I have some changes stashed locally which throw a custom error when the reply is already hijacked, I fail to see any reason why it might be necessary though which is why I haven't pushed them.

Reply:
I have been thinking about this for a while. How is this different from setting `reply.sent = true`? Could we just do that inside `hijack()` instead?

Reply:
I've experimented with `reply.sent = true` today and all tests in both repos pass except the [`onTimeout` hook test](https://github.com/AyoubElk/fastify-websocket/blob/websocket-fastifyrequest/test/hooks.js#L49) in `fastify-websocket`.

Upon inspection, it seems that the `hookIterator` function checks if the reply is sent, which causes the `onTimeout` hooks to be skipped:

```js
function hookIterator (fn, request, reply, next) {
  if (reply.sent === true) return undefined
  return fn(request, reply, next)
}
```

The current implementation I have makes use of `onTimeout` since I think it's still applicable in the context of websocket connections, do you agree with that?

Addtionally, I find that separating `reply[kReplyHijacked]` from the `reply[kReplySent]` makes it easier to reason about those parts of the codebase separately, but if you feel like it would be best to merge the two let me know and I'll try to figure out a solution for the issue described above.

Reply:
I would prefer to merge the two. Every property we add to `Request` or `Reply` increase the memory footprint of a request, slowing us down.

Reply:
> The current implementation I have makes use of onTimeout since I think it's still applicable in the context of websocket connections, do you agree with that?

I don't think it is a good API to have hijacked requests still trigger timeout hooks. The `onTimeout` hook means something specific in Fastify right now which is specific to the context of HTTP1 requests or HTTP2 sessions, and I think existing code is written with that in mind. If hijacked requests that become websockets still trigger the timeouts, then it gets hard to have a plugin encapsulation context that has both normal requests and websocket requests with a timeout that makes sense. You either want to keep it short and timeout normal requests, which would prematurely trrigger for websocket requests, or you want to keep it long and not timeout long lived websocket requests, which would allow erroneously long normal requests to go too long. I think timeouts for Fastify should only apply to requests that Fastify completes the processing for, and then `fastify-websocket` should have its own timeout mechanism, or that should be left up to consumers of it.

> I would prefer to merge the two. Every property we add to Request or Reply increase the memory footprint of a request, slowing us down.

You're ok with adding new stuff to the prototype though, right? `reply.hijack` can still be a thing?

Reply:
> I don't think it is a good API to have hijacked requests still trigger timeout hooks. The `onTimeout` hook means something specific in Fastify right now which is specific to the context of HTTP1 requests or HTTP2 sessions, and I think existing code is written with that in mind. If hijacked requests that become websockets still trigger the timeouts, then it gets hard to have a plugin encapsulation context that has both normal requests and websocket requests with a timeout that makes sense. You either want to keep it short and timeout normal requests, which would prematurely trrigger for websocket requests, or you want to keep it long and not timeout long lived websocket requests, which would allow erroneously long normal requests to go too long. I think timeouts for Fastify should only apply to requests that Fastify completes the processing for, and then `fastify-websocket` should have its own timeout mechanism, or that should be left up to consumers of it.
> 

Thanks! I was operating under the assumption that the meaning of `onTimeout` in fastify would extend to websocket connections as well, but this makes more sense to me.